pkgs/build-support/kernel/make-initrd-ng.nix

```diff
@@ -0,0 +1,79 @@
+  if lib.isFunction compressor then compressor
+  else if ! builtins.hasContext compressor && builtins.hasAttr compressor compressors then compressors.${compressor}.executable
+  else _: compressor
+, _compressorExecutable ? _compressorFunction pkgsBuildHost
+, _compressorName ? compressorName _compressorExecutable
+, _compressorMeta ? compressors.${_compressorName} or {}
+
+# List of arguments to pass to the compressor program, or null to use its defaults
+, compressorArgs ? null
+, _compressorArgsReal ? if compressorArgs == null then _compressorMeta.defaultArgs or [] else compressorArgs
+
+# Filename extension to use for the compressed initramfs. This is
+# included for clarity, but $out/initrd will always be a symlink to
+# the final image.
+# If this isn't guessed, you may want to complete the metadata above and send a PR :)
+, extension ? _compressorMeta.extension or
+    (throw "Unrecognised compressor ${_compressorName}, please specify filename extension")
+
+# List of { object = path_or_derivation; symlink = "/path"; }
+# The paths are copied into the initramfs in their nix store path
+# form, then linked at the root according to `symlink`.
+, contents
+
+# List of uncompressed cpio files to prepend to the initramfs. This
+# can be used to add files in specified paths without them becoming
+# symlinks to store paths.
+, prepend ? []
+
+# Whether to wrap the initramfs in a u-boot image.
+, makeUInitrd ? stdenvNoCC.hostPlatform.linux-kernel.target == "uImage"
+
+# If generating a u-boot image, the architecture to use. The default
+# guess may not align with u-boot's nomenclature correctly, so it can
+# be overridden.
+# See https://gitlab.denx.de/u-boot/u-boot/-/blob/9bfb567e5f1bfe7de8eb41f8c6d00f49d2b9a426/common/image.c#L81-106 for a list.
+, uInitrdArch ? stdenvNoCC.hostPlatform.linuxArch
+
+# The name of the compression, as recognised by u-boot.
+# See https://gitlab.denx.de/u-boot/u-boot/-/blob/9bfb567e5f1bfe7de8eb41f8c6d00f49d2b9a426/common/image.c#L195-204 for a list.
+# If this isn't guessed, you may want to complete the metadata above and send a PR :)
+, uInitrdCompression ? _compressorMeta.ubootName or
+    (throw "Unrecognised compressor ${_compressorName}, please specify uInitrdCompression")
+}: runCommand name {
+  compress = "${_compressorExecutable} ${lib.escapeShellArgs _compressorArgsReal}";
+  passthru = {
+    compressorExecutableFunction = _compressorFunction;
+    compressorArgs = _compressorArgsReal;
+  };
+
+  passAsFile = ["contents"];
+  contents = lib.concatMapStringsSep "\n" ({ object, symlink, ... }: "${object}\n${if symlink == null then "" else symlink}") contents + "\n";
+
+  nativeBuildInputs = [makeInitrdNGTool patchelf glibc cpio];
+} ''
+  mkdir ./root
+  make-initrd-ng "$contentsPath" ./root
+  mkdir "$out"
+  (cd root && find * .[^.*] -exec touch -h -d '@1' '{}' +)
+  (cd root && find * .[^.*] -print0 | sort -z | cpio -o -H newc -R +0:+0 --reproducible --null | eval -- $compress >> "$out/initrd")
```

Comment:
What do these find statements match? What is the benefit of using the wildcard `*` over `.` here? I think the regex could excludes `..`, which wouldn't be matched with `find . ` or `find root`.

Reply:
@mweinelt A ton of this file was just copied from the old makeInitrd. In the case of the `find` pipeline, I figured there must have been *some* reason for it, so I did it the same way instead of what I thought would have been more obvious.

Reply:
cc @lheckemann @joachifm @abbradar

Reply:
The biggest difference I would see is that if you `find .` you

1. get `.`, and
2. your name are proceeded with a `./`.

From a bit of testing, the only difference it makes to the archive is whether it contains a `.` entry as the leading `./` are stripped from the names. I don't know if this matters at all, but I suspect not unless it somehow messes with ownership or permissions of the root directory when extracting.

I would also guess that the `.[^.*]` is a mistake (it matches two letter names where the first is `.` and the second is not `.` or `*`). Most likely the intent was `.[^.]*` to get all hidden names (i.e., all `.*` names with the exception of `..`).

Any, if you want the same behaviour, from a `find .` variant, it would be (`%P` excludes the specified starting point)

```
find . -printf '%P\0'
```

It will generate a warning from `cpio` about ignoring the empty line (which is `.` with the `.` suppressed). Given `cpio` strips the leading `./`, another option that just suppresses `.` would be

```
find . -path . -o -print0
```